pf4j/src/main/java/org/pf4j/PluginState.java

```diff
@@ -57,7 +57,14 @@ public enum PluginState {
     /**
      * Plugin failed to start.
      */
-    FAILED("FAILED");
+    FAILED("FAILED"),
+
+    /**
+     * The plugin has been unloaded. After this event has been completed, the plugin's
+     * {@link ClassLoader} will be closed.
```

Comment:
I think the second sentence here is important. This even needs to be posted _before_ closing the `ClassLoader`, otherwise we prevent event listeners from doing some things that involve the classes in the plugin itself. I'm just not sure how to write a test for it :thinking: Do you have any ideas @decebals?

Reply:
I think that we can add a flag "closed" in `PluginClassLoader` and a method `isClosed():boolean`.  In the test method we will test the value returned by the `isClosed` method. In the future maybe we can use this flag to throw some exceptions in different methods with the error message "The class loader is already closed".
Or we can skip this test for the moment.
PluginClassLoader extends URLClassLoader and URLCLassLoader contains a [close](https://docs.oracle.com/javase/8/docs/api/java/net/URLClassLoader.html#close--) method with the below javadoc:

> Closes this URLClassLoader, so that it can no longer be used to load new classes or resources that are defined by this loader.

We need to investigate in the future more carefully this part with closed ClassLoader and the implications of this operation.

Reply:
Good idea :+1: I gave it a try, pushed it as a separate commit.

> We need to investigate in the future more carefully this part with closed ClassLoader and the implications of this operation.

This would not hurt. I don't know if it's a problem, but I think we are forced to make the `boolean closed` `PluginClassLoader` field `volatile`, so checking it in the overridden method comes at some cost. OTOH, the methods are already quite heavy, and should maybe not be called on super-hot paths.

It could also cause some inconsistencies, unless we override more of the methods from the superclass :thinking: But that is probably just a minor annoyance.